lib/nosedrum/storage.ex

```diff
@@ -1,75 +1,142 @@
+  @callback_type_map %{
+    pong: 1,
+    channel_message_with_source: 4,
+    deferred_channel_message_with_source: 5,
+    deferred_update_message: 6,
+    update_message: 7
+  }
 
-  In addition to subcommand names, the key `:default` can be specified by
-  the module. `:default` should be invoked when none of the subcommands in the
-  map match.
-  """
-  @type command_group ::
-          module
-          | %{optional(:default) => command_group, required(String.t()) => command_group}
+  @flag_map %{
+    ephemeral?: 64
+  }
+
+  @type command_scope :: :global | Guild.id() | [Guild.id()]
 
   @typedoc """
-  The "invocation path" of the command.
+  Defines a structure of commands, subcommands, subcommand groups, as
+  outlined in the [official documentation](https://discord.com/developers/docs/interactions/application-commands#subcommands-and-subcommand-groups).
 
-  The public-facing API of storage modules should use this in order
-  to allow users to identify the command they want to operate on.
+  **Note** that Discord only supports nesting 3 levels deep, like `command -> subcommand group -> subcommand`.
 
-  ## Usage
-  To identify a single command, use a single element list, such as `["echo"]`.
-  To identify a subcommand, use a pair, such as `["infraction", "search"]`.
-  To identify the default subcommand invoked when no matching subcommand is
-  found, specify the group name first, then `:default`, such as
-  `["tags", :default]`.
+  ## Example path:
+  ```elixir
+  %{
+    {"castle", MyApp.CastleCommand.description()} =>
+      %{
+        {"prepare", "Prepare the castle for an attack."} => [],
+        {"open", "Open up the castle for traders and visitors."} => [],
+        # ...
+      }
+  }
+  ```
+  """
+  @type application_command_path ::
+          %{
+            {group_name :: String.t(), group_desc :: String.t()} => [
+              application_command_path | [Nosedrum.ApplicationCommand.option()]
+            ]
+          }
+
+  @typedoc """
+  The name or pid of the Storage process.
   """
-  @type command_path :: [String.t() | :default, ...]
+  @type name_or_pid :: atom() | pid()
 
   @doc """
-  Look up a command group under the specified `name`.
+  Handle an Application Command invocation.
 
-  If the command was not found, `nil` should be returned.
+  This callback should be invoked upon receiving an interaction via the `:INTERACTION_CREATE` event.
+
+  ## Example using `Nosedrum.Storage.Dispatcher`:
+  ```elixir
+  # In your `Nostrum.Consumer` file:
+  def handle_event({:INTERACTION_CREATE, interaction, _ws_state}) do
+    IO.puts "Got interaction"
+    Nosedrum.Storage.Dispatcher.handle_interaction(interaction)
+  end
+  ```
   """
-  @callback lookup_command(name :: String.t(), storage :: reference) :: command_group | nil
+  @callback handle_interaction(interaction :: Interaction.t(), name_or_pid) ::
+              {:ok} | {:error, :unknown_command} | Nostrum.Api.error()
```

Comment:
I think
```suggestion
              {:ok} | {:error, :unknown_command, Nostrum.Api.error()}
```
is what we want here, isn't it?

Reply:
It looks like [`Nostrum.Api.error()`](https://hexdocs.pm/nostrum/Nostrum.Api.html#t:error/0) is an error tuple itself, though if you like the `|` in the tuple I think we could use `Nostrum.Api.ApiError.t()` as the 2nd element instead?

Reply:
Ahh, sorry, you are right. This is correct then 🙂